From 2a09fd62c23880ff306499ae03fe64628d82a23f Mon Sep 17 00:00:00 2001
From: Tim Gross <tgross@hashicorp.com>
Date: Thu, 7 May 2026 14:05:22 -0400
Subject: [PATCH] protect logmon from symlink attacks (#3957)

The logmon hook for each task creates two FIFOs in the alloc logs directory and
spawns a `logmon` subprocess that attaches to those FIFOs. When a task is
restarted, the `logmon` process is also recreated. There are two ways in which
this can be used to attack the host:

First, a malicious workload can use this to read arbitrary files from the host
file system by replacing the FIFO with a symlink. Inside the task's mount
namespace, the symlink points to a file inside the container (or nowhere). But
in the host's mount namespace, which is where `logmon` runs, the symlink points
to an arbitrary file on the host. The workload then restarts its own task. At
this point, `logmon` will be recreated and read through the symlink to the
target file, writing it to the task logs where they can be read by the
attacker. This allows exfiltration of arbitrary file contents from the host or
another task running on the same host.

Second, task drivers `FSIsolationImage` or `FSIsolationChroot` bind-mount the
`ALLOC_DIR` directory to the task. The logs directory containing both the FIFOs
and the log files are found in a subdirectory of this directory. During
investigation of the first issue, we discovered that by removing the
`/alloc/logs/` subdirectory and replacing it with a symlink, a task can cause
logmon to create the log files and FIFOs in arbitrary directories on the
host. The names of these files are only partially attacker controlled, but there
may be host configurations where this control over specific services on the host
or denial of service.

This changeset includes two fixes:
* Update the FIFO helper on Unix to use `*os.Root` to prevent following symlinks
  outside the logs directory, traversing file system boundaries, or accessing
  special files.
  * On Linux and BSDs that support `mkfifoat` (i.e. not macOS) we'll also use
    that syscall specifically when creating the FIFO.
  * On Windows, we don't need this because named pipes are in their own kernel
    namespace and not present in the file system to be manipulated by the task.
* Mount the `/alloc/logs` subdirectory as a separate read-only bind-mount to the
  task. This prevents the task from writing directly to the logs directory
  entirely, except in cases where the task has an unconstrained filesystem (as
  in `raw_exec`).

Fixes: https://hashicorp.atlassian.net/browse/SECVULN-42948
Ref: https://hashicorp.atlassian.net/browse/NMD-1432
---
 .changelog/27918.txt           |  7 ++++
 client/allocdir/fs_darwin.go   |  2 +-
 client/allocdir/fs_freebsd.go  |  2 +-
 client/allocdir/fs_linux.go    |  5 ++-
 client/allocdir/fs_netbsd.go   |  2 +-
 client/allocdir/fs_solaris.go  |  2 +-
 client/allocdir/fs_windows.go  |  2 +-
 client/allocdir/task_dir.go    | 12 ++++++-
 client/lib/fifo/doc.go         | 12 ++++---
 client/lib/fifo/fifo_test.go   |  8 +++--
 client/lib/fifo/fifo_unix.go   | 58 ++++++++++++++++++++++++++--------
 client/lib/fifo/mkfifo_unix.go | 21 ++++++++++++
 client/lib/fifo/mkfifoat.go    | 39 +++++++++++++++++++++++
 drivers/docker/driver.go       |  7 +++-
 drivers/docker/driver_test.go  |  2 ++
 15 files changed, 154 insertions(+), 27 deletions(-)
 create mode 100644 .changelog/27918.txt
 create mode 100644 client/lib/fifo/mkfifo_unix.go
 create mode 100644 client/lib/fifo/mkfifoat.go

diff --git a/.changelog/27918.txt b/.changelog/27918.txt
new file mode 100644
index 00000000000..43638fd1bf6
--- /dev/null
+++ b/.changelog/27918.txt
@@ -0,0 +1,7 @@
+```release-note:security
+logging: Protect logging FIFO from symlink swap attacks (CVE-2026-6959)
+```
+
+```release-note:breaking-change
+logging: The allocation logs directory is bind-mounted read-only for task drivers that support with filesystem isolation
+```
diff --git a/client/allocdir/fs_darwin.go b/client/allocdir/fs_darwin.go
index 9c3900b80f3..2ad60ac286f 100644
--- a/client/allocdir/fs_darwin.go
+++ b/client/allocdir/fs_darwin.go
@@ -9,7 +9,7 @@ import (
 )
 
 // linkDir hardlinks src to dst. The src and dst must be on the same filesystem.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, _ bool) error {
 	return syscall.Link(src, dst)
 }
 
diff --git a/client/allocdir/fs_freebsd.go b/client/allocdir/fs_freebsd.go
index 9c3900b80f3..2ad60ac286f 100644
--- a/client/allocdir/fs_freebsd.go
+++ b/client/allocdir/fs_freebsd.go
@@ -9,7 +9,7 @@ import (
 )
 
 // linkDir hardlinks src to dst. The src and dst must be on the same filesystem.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, _ bool) error {
 	return syscall.Link(src, dst)
 }
 
diff --git a/client/allocdir/fs_linux.go b/client/allocdir/fs_linux.go
index 445e249a209..f4713c1a4ef 100644
--- a/client/allocdir/fs_linux.go
+++ b/client/allocdir/fs_linux.go
@@ -22,11 +22,14 @@ const (
 
 // linkDir bind mounts src to dst as Linux doesn't support hardlinking
 // directories.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, ro bool) error {
 	if err := os.MkdirAll(dst, fileMode777); err != nil {
 		return err
 	}
 
+	if ro {
+		return syscall.Mount(src, dst, "", syscall.MS_BIND|syscall.MS_RDONLY, "")
+	}
 	return syscall.Mount(src, dst, "", syscall.MS_BIND, "")
 }
 
diff --git a/client/allocdir/fs_netbsd.go b/client/allocdir/fs_netbsd.go
index 9c3900b80f3..2ad60ac286f 100644
--- a/client/allocdir/fs_netbsd.go
+++ b/client/allocdir/fs_netbsd.go
@@ -9,7 +9,7 @@ import (
 )
 
 // linkDir hardlinks src to dst. The src and dst must be on the same filesystem.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, _ bool) error {
 	return syscall.Link(src, dst)
 }
 
diff --git a/client/allocdir/fs_solaris.go b/client/allocdir/fs_solaris.go
index 1cc7a9f2d8a..98894714606 100644
--- a/client/allocdir/fs_solaris.go
+++ b/client/allocdir/fs_solaris.go
@@ -9,7 +9,7 @@ import (
 )
 
 // linkDir hardlinks src to dst. The src and dst must be on the same filesystem.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, _ bool) error {
 	return syscall.Link(src, dst)
 }
 
diff --git a/client/allocdir/fs_windows.go b/client/allocdir/fs_windows.go
index c793a11a5ff..feb5027c9dc 100644
--- a/client/allocdir/fs_windows.go
+++ b/client/allocdir/fs_windows.go
@@ -28,7 +28,7 @@ func linkOrCopy(src, dst string, uid, gid int, perm os.FileMode) error {
 }
 
 // The windows version does nothing currently.
-func linkDir(src, dst string) error {
+func linkDir(src, dst string, _ bool) error {
 	return nil
 }
 
diff --git a/client/allocdir/task_dir.go b/client/allocdir/task_dir.go
index c515466ce52..93ac3625a87 100644
--- a/client/allocdir/task_dir.go
+++ b/client/allocdir/task_dir.go
@@ -153,9 +153,13 @@ func (t *TaskDir) Build(fsi fsisolation.Mode, chroot map[string]string, username
 		// If the path doesn't exist OR it exists and is empty, link it
 		empty, _ := pathEmpty(t.SharedTaskDir)
 		if !pathExists(t.SharedTaskDir) || empty {
-			if err := linkDir(t.SharedAllocDir, t.SharedTaskDir); err != nil {
+			if err := linkDir(t.SharedAllocDir, t.SharedTaskDir, false); err != nil {
 				return fmt.Errorf("Failed to mount shared directory for task: %w", err)
 			}
+			if err := linkDir(t.LogDir, filepath.Join(t.SharedTaskDir, "logs"), true); err != nil {
+				return fmt.Errorf("Failed to mount shared directory for task: %w", err)
+			}
+
 		}
 	}
 
@@ -326,6 +330,12 @@ func (t *TaskDir) Unmount() error {
 
 	// Check if the directory has the shared alloc mounted.
 	if pathExists(t.SharedTaskDir) {
+		if err := unlinkDir(filepath.Join(t.SharedTaskDir, "logs")); err != nil {
+			mErr = multierror.Append(mErr,
+				fmt.Errorf("failed to unmount logs dir %q: %w",
+					filepath.Join(t.SharedTaskDir, "logs"), err))
+		}
+
 		if err := unlinkDir(t.SharedTaskDir); err != nil {
 			mErr = multierror.Append(mErr,
 				fmt.Errorf("failed to unmount shared alloc dir %q: %w", t.SharedTaskDir, err))
diff --git a/client/lib/fifo/doc.go b/client/lib/fifo/doc.go
index 795a1d55606..86564ddf728 100644
--- a/client/lib/fifo/doc.go
+++ b/client/lib/fifo/doc.go
@@ -3,9 +3,13 @@
 
 /*
 Package fifo implements functions to create and open a fifo for inter-process
-communication in an OS agnostic way. A few assumptions should be made when
-using this package. First, New() must always be called before Open(). Second
-Open() returns an io.ReadWriteCloser that is only connected with the
-io.ReadWriteCloser returned from New().
+communication in an OS agnostic way. A few assumptions should be made when using
+this package. First, New() must always be called before Open(). Second Open()
+returns an io.ReadWriteCloser that is only connected with the io.ReadWriteCloser
+returned from New().
+
+On Unix, all exported functions use os.Root under the hood to avoid chasing
+symlinks out of their parent directory. On Windows, this is unnecessary because
+named pipes exist in their own namespace and not the filesystem.
 */
 package fifo
diff --git a/client/lib/fifo/fifo_test.go b/client/lib/fifo/fifo_test.go
index bd0cca32cd2..3e837b31f09 100644
--- a/client/lib/fifo/fifo_test.go
+++ b/client/lib/fifo/fifo_test.go
@@ -22,10 +22,12 @@ func TestFIFO(t *testing.T) {
 	require := require.New(t)
 	var path string
 
+	rootDir := t.TempDir()
+
 	if runtime.GOOS == "windows" {
 		path = "//./pipe/fifo"
 	} else {
-		path = filepath.Join(t.TempDir(), "fifo")
+		path = filepath.Join(rootDir, "fifo")
 	}
 
 	readerOpenFn, err := CreateAndRead(path)
@@ -82,10 +84,12 @@ func TestWriteClose(t *testing.T) {
 	require := require.New(t)
 	var path string
 
+	rootDir := t.TempDir()
+
 	if runtime.GOOS == "windows" {
 		path = "//./pipe/" + uuid.Generate()[:4]
 	} else {
-		path = filepath.Join(t.TempDir(), "fifo")
+		path = filepath.Join(rootDir, "fifo")
 	}
 
 	readerOpenFn, err := CreateAndRead(path)
diff --git a/client/lib/fifo/fifo_unix.go b/client/lib/fifo/fifo_unix.go
index 41dc0312c8c..fb0917c0420 100644
--- a/client/lib/fifo/fifo_unix.go
+++ b/client/lib/fifo/fifo_unix.go
@@ -2,7 +2,6 @@
 // SPDX-License-Identifier: BUSL-1.1
 
 //go:build !windows
-// +build !windows
 
 package fifo
 
@@ -10,19 +9,19 @@ import (
 	"fmt"
 	"io"
 	"os"
-
-	"golang.org/x/sys/unix"
+	"path/filepath"
 )
 
-// CreateAndRead creates a fifo at the given path, and returns an open function for reading.
-// For compatibility with windows, the fifo must not exist already.
+// CreateAndRead creates a fifo at the given path, and returns an open function
+// for reading. For compatibility with windows, the fifo must not exist
+// already.
 //
 // It returns a reader open function that may block until a writer opens
 // so it's advised to run it in a goroutine different from reader goroutine
 func CreateAndRead(path string) (func() (io.ReadCloser, error), error) {
 	// create first
 	if err := mkfifo(path, 0600); err != nil {
-		return nil, fmt.Errorf("error creating fifo %v: %v", path, err)
+		return nil, fmt.Errorf("error creating fifo %v: %w", path, err)
 	}
 
 	return func() (io.ReadCloser, error) {
@@ -31,17 +30,54 @@ func CreateAndRead(path string) (func() (io.ReadCloser, error), error) {
 }
 
 func OpenReader(path string) (io.ReadCloser, error) {
-	return os.OpenFile(path, unix.O_RDONLY, os.ModeNamedPipe)
+	dir := filepath.Dir(path)
+	base := filepath.Base(path)
+
+	root, err := os.OpenRoot(dir)
+	if err != nil {
+		return nil, fmt.Errorf("error opening fifo parent directory %q: %w", dir, err)
+	}
+	defer root.Close()
+
+	// also uses O_NOFOLLOW under the hood
+	f, err := root.OpenFile(base, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("error opening reader at %s: %w", path, err)
+	}
+	return f, nil
 }
 
 // OpenWriter opens a fifo file for writer, assuming it already exists, returns io.WriteCloser
 func OpenWriter(path string) (io.WriteCloser, error) {
-	return os.OpenFile(path, unix.O_WRONLY, os.ModeNamedPipe)
+	dir := filepath.Dir(path)
+	base := filepath.Base(path)
+
+	root, err := os.OpenRoot(dir)
+	if err != nil {
+		return nil, fmt.Errorf("error opening fifo parent directory %q: %w", dir, err)
+	}
+	defer root.Close()
+
+	// also uses O_NOFOLLOW under the hood
+	f, err := root.OpenFile(base, os.O_WRONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("error opening writer at %s: %w", path, err)
+	}
+	return f, nil
 }
 
 // Remove a fifo that already exists at a given path
 func Remove(path string) error {
-	return os.Remove(path)
+	dir := filepath.Dir(path)
+	base := filepath.Base(path)
+
+	root, err := os.OpenRoot(dir)
+	if err != nil {
+		return fmt.Errorf("error opening root dir %q: %w", dir, err)
+	}
+	defer root.Close()
+
+	return root.Remove(base)
 }
 
 func IsClosedErr(err error) bool {
@@ -51,7 +87,3 @@ func IsClosedErr(err error) bool {
 	}
 	return false
 }
-
-func mkfifo(path string, mode uint32) (err error) {
-	return unix.Mkfifo(path, mode)
-}
diff --git a/client/lib/fifo/mkfifo_unix.go b/client/lib/fifo/mkfifo_unix.go
new file mode 100644
index 00000000000..3158ff0c197
--- /dev/null
+++ b/client/lib/fifo/mkfifo_unix.go
@@ -0,0 +1,21 @@
+// Copyright IBM Corp. 2015, 2025
+// SPDX-License-Identifier: BUSL-1.1
+
+//go:build !linux && !freebsd && !netbsd && !openbsd && !windows
+
+package fifo
+
+import (
+	"fmt"
+
+	"golang.org/x/sys/unix"
+)
+
+func mkfifo(path string, mode uint32) (err error) {
+	// macOS doesn't support mkfifoat
+	err = unix.Mkfifo(path, mode)
+	if err != nil {
+		return fmt.Errorf("error creating fifo: %w", err)
+	}
+	return nil
+}
diff --git a/client/lib/fifo/mkfifoat.go b/client/lib/fifo/mkfifoat.go
new file mode 100644
index 00000000000..049626c8df8
--- /dev/null
+++ b/client/lib/fifo/mkfifoat.go
@@ -0,0 +1,39 @@
+// Copyright IBM Corp. 2015, 2025
+// SPDX-License-Identifier: BUSL-1.1
+
+//go:build linux || freebsd || netbsd || openbsd
+
+package fifo
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+func mkfifo(path string, mode uint32) (err error) {
+	dir := filepath.Dir(path)
+	base := filepath.Base(path)
+
+	root, err := os.OpenRoot(dir)
+	if err != nil {
+		return fmt.Errorf("error opening fifo parent directory %q: %v", dir, err)
+	}
+	defer root.Close()
+
+	parent, err := root.Open(".")
+	if err != nil {
+		return fmt.Errorf("error getting file handle to fifo parent directory %q: %v", dir, err)
+	}
+	defer parent.Close()
+
+	// os.Root doesn't support creating a FIFO, so we need to drop to the
+	// syscall and grab the parent's FD
+	err = unix.Mkfifoat(int(parent.Fd()), base, mode)
+	if err != nil {
+		return fmt.Errorf("error creating fifo: %w", err)
+	}
+	return nil
+}
diff --git a/drivers/docker/driver.go b/drivers/docker/driver.go
index c294ca93ce1..b6bccf79eb0 100644
--- a/drivers/docker/driver.go
+++ b/drivers/docker/driver.go
@@ -769,15 +769,20 @@ func (d *Driver) containerBinds(task *drivers.TaskConfig, driverConfig *TaskConf
 	allocDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SharedAllocDir, task.Env[taskenv.AllocDir])
 	taskLocalBind := fmt.Sprintf("%s:%s", task.TaskDir().LocalDir, task.Env[taskenv.TaskLocalDir])
 	secretDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SecretsDir, task.Env[taskenv.SecretsDir])
+	selinuxLabel := d.config.Volumes.SelinuxLabel
+
 	binds := []string{allocDirBind, taskLocalBind, secretDirBind}
 
-	selinuxLabel := d.config.Volumes.SelinuxLabel
+	logsROFlag := "ro"
 	if selinuxLabel != "" {
 		// Apply SELinux Label to each built-in bind
 		for i := range binds {
 			binds[i] = fmt.Sprintf("%s:%s", binds[i], selinuxLabel)
 		}
+		logsROFlag = "ro," + selinuxLabel
 	}
+	allocLogsDirBind := fmt.Sprintf("%s/logs:%s/logs:%s", task.TaskDir().SharedAllocDir, task.Env[taskenv.AllocDir], logsROFlag)
+	binds = append(binds, allocLogsDirBind)
 
 	for _, userbind := range driverConfig.Volumes {
 		// This assumes host OS = docker container OS.
diff --git a/drivers/docker/driver_test.go b/drivers/docker/driver_test.go
index df4dfaea956..a6b351f36f1 100644
--- a/drivers/docker/driver_test.go
+++ b/drivers/docker/driver_test.go
@@ -1392,6 +1392,7 @@ func TestDockerDriver_CreateContainerConfig_Mounts(t *testing.T) {
 		"alloc:/alloc:z",
 		"redis-demo/local:/local:z",
 		"redis-demo/secrets:/secrets:z",
+		"alloc/logs:/alloc/logs:ro,z",
 		"/etc/ssl/certs:/etc/ssl/certs:ro,z",
 		"/var/www:/srv/www:z",
 	}, cc.Host.Binds)
@@ -1469,6 +1470,7 @@ func TestDockerDriver_CreateContainerConfig_Mounts_Windows(t *testing.T) {
 		`alloc:c:/alloc`,
 		`redis-demo\local:c:/local`,
 		`redis-demo\secrets:c:/secrets`,
+		`alloc/logs:c:/alloc/logs:ro`,
 		`c:\etc\ssl\certs:c:/etc/ssl/certs`,
 		`c:\var\www:c:/srv/www`,
 	}, cc.Host.Binds)
